nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/ConsumeAMQP.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.io.OutputStreamCallback;
+import org.apache.nifi.processor.util.StandardValidators;
+
+import com.rabbitmq.client.AMQP.BasicProperties;
+import com.rabbitmq.client.GetResponse;
+
+/**
+ * Consuming AMQP processor which upon each invocation of
+ * {@link #onTrigger(ProcessContext, ProcessSession)} method will construct a
+ * {@link FlowFile} containing the body of the consumed AMQP message and AMQP
+ * properties that came with message which are added to a {@link FlowFile} as
+ * attributes.
+ */
+@Tags({ "amqp", "rabbit", "get", "message", "receive", "consume" })
+@InputRequirement(Requirement.INPUT_FORBIDDEN)
+@CapabilityDescription("Creates a AMQP Message from the contents of a FlowFile and sends the message to an AMQP Server")
+public class ConsumeAMQP extends AbstractAMQPProcessor<AMQPConsumer> {
+
+    public static final PropertyDescriptor QUEUE = new PropertyDescriptor.Builder()
+            .name("Queue")
+            .description("Source queue")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("All FlowFiles that are received from the AMQP queue are routed to this relationship")
+            .build();
+
+    private final static List<PropertyDescriptor> propertyDescriptors;
+
+    private final static Set<Relationship> relationships;
+
+    /*
+     * Will ensure that the list of property descriptors is build only once.
+     * Will also create a Set of relationships
+     */
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.add(QUEUE);
+        _propertyDescriptors.addAll(descriptors);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_SUCCESS);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
+
+    /**
+     * Will construct a {@link FlowFile} containing the body of the consumed
+     * AMQP message (if {@link GetResponse} returned by {@link AMQPConsumer} is
+     * not null) and AMQP properties that came with message which are added to a
+     * {@link FlowFile} as attributes, transferring {@link FlowFile} to
+     * 'success' {@link Relationship}.
+     */
+    @Override
+    protected void rendezvousWithAmqp(ProcessContext context, ProcessSession processSession) throws ProcessException {
+        final GetResponse response = this.targetResource.consume();
+        if (response != null){
+            FlowFile flowFile = processSession.create();
+            flowFile = processSession.append(flowFile, new OutputStreamCallback() {
```

Comment:
I would change this to processSession.write - because the FlowFile has no content, the result will be the same (i.e., they have the same semantics) but the internal workings will actually be a little more efficient (and the code would be more consistent with what other processors do) if this is session.write